libraries/AC_PID/AC_PID_1D.h

```diff
@@ -0,0 +1,112 @@
+public:
+
+    // Constructor for PID
+    AC_PID_1D(float initial_p, float initial_i, float initial_d, float initial_ff, float initial_imax, float initial_filt_hz, float initial_filt_d_hz, float dt);
+
+    // set_dt - set time step in seconds
+    void set_dt(float dt);
+
+    // set target and measured inputs to PID controller and calculate outputs
+    // target and error are filtered
+    // the derivative is then calculated and filtered
+    // the integral is then updated based on the setting of the limit flag
+    float update_all(float target, float measurement, bool limit = false);
+    float update_all(float target, float measurement, bool limit_neg, bool limit_pos);
+
+    // update the integral
+    // if the limit flags are set the integral is only allowed to shrink
+    void update_i(bool limit_neg, bool limit_pos);
+
+    // get results from pid controller
+    float get_p() const;
+    float get_i() const;
+    float get_d() const;
+    float get_ff();
+    float get_error() const {return _error;}
+
+    // reset the integrator
+    void reset_I();
+
+    // input and D term filter will be reset to the next value provided to set_input()
+    void reset_filter() { _reset_filter = true; }
+
+    // load gain from eeprom
+    void load_gains();
+
+    // save gain to eeprom
+    void save_gains();
+
+    /// operator function call for easy initialisation
+    void operator()(float p_val, float i_val, float d_val, float ff_val, float imax_val, float input_filt_E_hz, float input_filt_D_hz, float dt);
+
+    // get accessors
+    AP_Float &kP() { return _kp; }
+    const AP_Float &kP() const { return _kp; }
+    float kI() const { return _ki; }
+    float kD() const { return _kd; }
+    float ff() const { return _kff;}
+    float imax() const { return _kimax.get(); }
+    float filt_E_hz() const { return _filt_E_hz; }
+    float filt_D_hz() const { return _filt_D_hz; }
+    float get_filt_E_alpha() const;
+    float get_filt_D_alpha() const;
+
+    // set accessors
+    void kP(const float v) { _kp.set(v); }
+    void kI(const float v) { _ki.set(v); }
+    void kD(const float v) { _kd.set(v); }
+    void ff(const float v) { _kff.set(v); }
+    void imax(const float v) { _kimax.set(fabsf(v)); }
+    void filt_E_hz(const float v);
+    void filt_D_hz(const float v);
+
+    // integrator setting functions
+    void set_integrator(float target, float measurement, float i);
+    void set_integrator(float error, float i);
+    void set_integrator(float i);
+
+    const AP_Logger::PID_Info& get_pid_info(void) const { return _pid_info; }
+
+    // parameter var table
+    static const struct AP_Param::GroupInfo        var_info[];
+
+protected:
+
+    // parameters
+    AP_Float _kp;
+    AP_Float _ki;
+    AP_Float _kd;
+    AP_Float _kff;
+    AP_Float _kimax;
+    AP_Float _filt_E_hz;         // PID error filter frequency in Hz
+    AP_Float _filt_D_hz;         // PID derivative filter frequency in Hz
+
+    // internal variables
```

Comment:
So... why aren't they private?

Reply:
replaced protected with private for parameters and members